flake.nix

```diff
@@ -13,56 +13,112 @@
 
   inputs = {
     nixpkgs.url = "github:NixOS/nixpkgs/nixos-unstable-small";
-    flake-utils.url = "github:numtide/flake-utils";
-
-    flake-compat = {
-      url = "github:edolstra/flake-compat";
-      flake = false;
-    };
-
-    haskell-flake-utils = {
-      url = "github:ivanovs-4/haskell-flake-utils";
-      inputs.flake-utils.follows = "flake-utils";
-    };
   };
 
-outputs = { self, nixpkgs, flake-utils, haskell-flake-utils, flake-compat, ... }:
-  haskell-flake-utils.lib.simpleCabalProject2flake {
-    inherit self nixpkgs;
+  outputs = inputs:
+    with builtins;
+    let
+      lib = inputs.nixpkgs.lib;
+      pnames = [
+        "automaton"
+        "rhine"
+        "rhine-bayes"
+        "rhine-examples"
+        "rhine-gloss"
+        "rhine-terminal"
+      ];
+      supportedGhcs = [ "ghc92" "ghc94" "ghc96" "ghc98" ];
+      libPnames = filter (pname: pname != "rhine-examples") pnames;
+      hpsFor = pkgs:
+        lib.filterAttrs (name: _: elem name supportedGhcs) pkgs.haskell.packages
+        // { default = pkgs.haskellPackages; };
+      rhinePackages = hfinal: hprev:
+        lib.genAttrs pnames (pname: hfinal.callCabal2nix pname ./${pname} { });
+      overlay = final: prev:
+        let hps = hpsFor final; in {
+          haskell = prev.haskell // {
+            packageOverrides = with prev.haskell.lib.compose; lib.composeManyExtensions [
+              prev.haskell.packageOverrides
+              rhinePackages
+              (hfinal: hprev: {
+                monad-bayes = markUnbroken hprev.monad-bayes;
+                monad-schedule = hprev.callHackageDirect
+                  {
+                    pkg = "monad-schedule";
+                    ver = "0.2";
+                    sha256 = "sha256-Z9lAxkvJDH9aQZd65bGOQI3EGH7oSAhK0nuBKULgiCE=";
+                  }
+                  { };
+                time-domain = hprev.callHackageDirect
+                  {
+                    pkg = "time-domain";
+                    ver = "0.1.0.5";
+                    sha256 = "sha256-llDBQuU5ez/0MiOIMH97P4BQhFDyPfTMWinq1wJrDGI=";
+                  }
+                  { };
+              })
+              (hfinal: hprev: lib.optionalAttrs prev.stdenv.isDarwin {
+                monad-schedule = dontCheck hprev.monad-schedule;
+              })
+              (hfinal: hprev: lib.optionalAttrs (lib.versionOlder hprev.ghc.version "9.4") {
+                time-domain = doJailbreak hprev.time-domain;
+              })
+            ];
+          };
+          rhine-bin = prev.buildEnv {
+            name = "rhine-bin";
+            paths = map (pname: hps.default.${pname}) pnames;
+            pathsToLink = [ "/bin" ];
+          };
```

Comment:
I'm not sure what this is for. The executables should be built by `rhine-lib` as well, right?

Reply:
Ah, but in `/lib` there is a directory structure that disambiguates the ghc version. Since this doesn't exist in `/bin` we need to chose a GHC version here, hence the different treatment (if we want the executables as a build output).

Reply:
Let's leave it like this for now. I'm wondering whether there is value in having a structure `/bin/ghc98/executable1, /bin/ghc96/executable2`, and so on.